src/bootstrap/llvm.rs

```diff
@@ -834,6 +834,30 @@ impl Step for Lld {
             }
         }
 
+        // LLD is built as an LLVM tool, but is distributed outside of the `llvm-tools` component,
+        // which impacts where it expects to find LLVM's shared library. This causes #80703.
+        //
+        // LLD is distributed at "$root/lib/rustlib/$host/bin/rust-lld", but the `libLLVM-*.so` it
+        // needs is distributed at "$root/lib". The default rpath of "$ORIGIN/../lib" points at the
+        // lib path for LLVM tools, not the one for rust binaries.
+        //
+        // (The `llvm-tools` component copies the .so there for the other tools, and with that
+        // component installed, one can successfully invoke `rust-lld` directly without rustup's
+        // `LD_LIBRARY_PATH` overrides)
+        //
+        if builder.config.rust_rpath
+            && builder.config.llvm_link_shared()
+            && target.contains("linux")
```

Comment:
I think we probably want a slightly more complex condition rather than just rust_rpath here, at least conditional on the specific target we're building for?

See https://github.com/rust-lang/rust/blob/1e17cef9e29961636db63730784caea4b7d47ff5/src/bootstrap/builder.rs#L1629 for what we do for regular rustc/std/etc.

I'm also wondering whether we need this on Apple targets.

Reply:
I believe we don’t use link-shared on mac ? (at least not on my M1)

Reply:
> I think we probably want a slightly more complex condition rather than just rust_rpath here, at least conditional on the specific target we're building for?

Yes absolutely, good call. I've pushed a commit to add the more complex condition.